File: src/Olivine/Core/Pass/Unroll.hs
-- | Loops written out turn by turn, where the turns are countable and few.
--
-- What a loop costs beyond the work in it is a test and a branch per turn, and
-- for a body of three instructions that is most of what runs.  Writing the
-- turns out one after another takes the test and the branch away, and it does
-- something worth more than either: it puts every turn in one block, where the
-- passes below can see them together.  Two turns reading the same address are
-- two loads one walk can answer; a counter that was a different number each
-- time round is a different constant in each copy, and an address computed from
-- it is an address folding settles.
--
-- __The counter's value is written down, not worked out again.__  Each copy
-- opens with an assignment saying what the loop's own locals hold at the top of
-- that turn, which "Olivine.Core.Induction" had to know in order to count the
-- turns at all.  Without it a copy is the same work as the turn it copies —
-- the counter still computed from the last one — and the pass would trade a
-- branch for a block four times the size.  With it the copies are what folding
-- reads through, and this is the pass that makes constants of a loop's
-- arithmetic rather than the pass that removes a branch.
--
-- The assignments say nothing that was not already true: a local holds what the
-- interpreter says it holds at the top of that turn, or it is not mentioned.
--
-- __What is written out is bounded twice__, by 'turnLimit' on how many turns
-- there may be and by 'bodyBudget' on how much may be written out in total.
-- Unrolling is the one transformation here that makes a function bigger on
-- purpose, so the bound is what says how much bigger; and unlike the other
-- bounds in the optimizer it is also what makes an analysis terminate, since
-- counting the turns of a loop means running them.
--
-- __The loop has to be one block that branches to itself.__  That is what a
-- small counted loop is by the time this runs — rotation puts the test at the
-- bottom and block merging makes the body and the test one block — and it is
-- what makes writing the turns out a matter of putting the copies in a row.  A
-- loop of several blocks would want its paths copied and its labels reissued,
-- which is a different pass; nothing in the corpus wants it yet.
module Olivine.Core.Pass.Unroll
  ( unrollLoops
  , turnLimit
  , bodyBudget
  ) where

import Data.List (find)
import Data.Map.Strict (Map)
import Data.Map.Strict qualified as Map
import Data.Set qualified as Set

import Olivine.Core.Induction (Turn (..), turnsOf)
import Olivine.Core.Instruction
import Olivine.Core.Loops (Loop (..), loopsOf)
import Olivine.Core.Program
import Olivine.Syntax.Name (Name)
import Olivine.Syntax.Type (Type)
import Olivine.Syntax.Value (TypedValue (..))

-- | How many turns a loop may make and still be written out.
--
-- The number is what a body of a few instructions is worth writing out that
-- many copies of, and it is also how far 'turnsOf' is asked to interpret,
-- which is the whole cost of asking.
turnLimit :: Int
turnLimit = 8

-- | How many instructions all the copies together may come to.
--
-- Counted on the loop as written rather than on what is left after folding,
-- because what folding will make of a copy is not known until it has run.  The
-- copies are what pays for the branch and the counter, so the budget is the
-- point at which a bigger function stops being a faster one.
bodyBudget :: Int
bodyBudget = 40

unrollLoops :: Program -> Program
unrollLoops program =
  program {programEntries = map entry (programEntries program)}
  where
    types = namedTypes program
    entry (EFunction f) = EFunction (settle types f)
    entry retained = retained

-- | Writing one loop out may leave a block that is another loop's whole body,
-- so this runs until a sweep finds nothing.  A sweep that writes a loop out
-- removes a back edge and adds none, so the sweeps stop.
settle :: Map Name Type -> Function -> Function
settle types f = case [written | loop <- loopsOf f, Just written <- [unroll types f loop]] of
  written : _ -> settle types written
  [] -> f

-- | One loop written out, if it is one of the loops this is for.
unroll :: Map Name Type -> Function -> Loop -> Maybe Function
unroll types f loop = do
  body <- find ((== header) . blockLabel) (functionBlocks f)
  turns <- turnsOf turnLimit f loop
  leaving <- wayOut body
  let instructions = blockInstructions body
  -- The budget is against what will be written, which is the body once per
  -- turn — but only the part of it that costs anything.  An assignment is not
  -- an instruction in the output: reconstruction takes every one of them away,
  -- and a body a front end wrote is mostly the assignments promotion left where
  -- its loads and stores were.  Counting those would refuse to write out a loop
  -- whose real body is five instructions because it is written in thirteen.
  guardOn (length turns * length (filter costly instructions) <= bodyBudget)
  let Local next = nextLocal f
      copies =
        concat
          [ known turn <> namedApart types (Local (next + n * length instructions)) instructions
          | (n, turn) <- zip [0 ..] turns
          ]
  pure
    f
      { functionBlocks =
          [ if blockLabel b == header
              then
                b
                  { blockInstructions = copies
                  , -- The loop is gone, so what it promised about itself goes
                    -- with it: a @!llvm.loop@ left on a branch that closes
                    -- nothing is a promise about a loop nobody can find.
                    blockTerminator = Terminator (Br leaving) []
                  }
              else b
          | b <- functionBlocks f
          ]
      }
  where
    header = loopHeader loop

    -- Where control goes when the loop is done, which is the way out of the
    -- block that is not back into it.
    wayOut body = case terminatorTransfer (blockTerminator body) of
      CondBr _ takenIf takenElse
        | takenIf == header, takenElse /= header -> Just takenElse
        | takenElse == header, takenIf /= header -> Just takenIf
      _ -> Nothing

    -- What the turn is known to hold, said as assignments at the top of the
    -- copy.  A type is wanted for each and the local's own type is what the
    -- instruction that assigns it in the body gives.
    known turn =
      [ Instruction (Just name) (OAssign (TypedValue t value)) []
      | (name, value) <- Map.toList (turnKnown turn)
      , Just t <- [typeOf name]
      ]

    typeOf name =
      case [ resultType types (instructionOperation i)
           | b <- functionBlocks f
           , Set.member (blockLabel b) (loopBody loop)
           , i <- blockInstructions b
           , instructionResult i == Just name
           ] of
        t : _ -> Just t
        [] -> Nothing

    costly i = case instructionOperation i of
      OAssign _ -> False
      _ -> True

    guardOn condition = if condition then Just () else Nothing

File: test/c/unroll.c
/* Loops whose turns can be counted, and loops whose turns cannot.
 *
 * Written in pairs wherever there is a pair to write: the same loop with a
 * bound the compiler knows and with one only the caller knows, so that what
 * the counting buys is the difference between the two rather than a number on
 * its own.  Every function is read back through what it computed, so a turn
 * written out wrongly — the counter a turn behind, a store to the wrong slot,
 * a body run once too often — shows up as a different number and not as a
 * shorter file. */

#include <stddef.h>

/* The plainest one: four turns, a constant bound, a body of two instructions.
   Every index is a constant once the turns are written out, so the address
   arithmetic goes with them. */
int total4(const int *p) {
  int s = 0;
  for (int i = 0; i < 4; i++)
    s += p[i];
  return s;
}

/* The same loop with a bound the caller supplies.  Nothing here is countable
   and the loop stays a loop: this is the one that says the pass is deciding
   rather than unrolling whatever it is given. */
int total_n(const int *p, int n) {
  int s = 0;
  for (int i = 0; i < n; i++)
    s += p[i];
  return s;
}

/* A step that is not one.  There is no rule here for what the closed form of
   this recurrence is, and none is wanted: the turns are counted by running
   them, so a step of three costs exactly what a step of one costs. */
int every_third(const int *p) {
  int s = 0;
  for (int i = 0; i < 12; i += 3)
    s += p[i];
  return s;
}

/* Counting down rather than up, which is the other comparison and the other
   sign of step. */
int count_back(const int *p) {
  int s = 0;
  for (int i = 4; i > 0; i--)
    s += p[i - 1];
  return s;
}

/* Storing rather than loading.  Each turn writes a different slot from a
   different value, so this is the one that fails loudly if the turns are
   written out sharing what they should not: all four stores landing on one
   slot is what a reconstruction that finds an instruction by value rather than
   by position does. */
void fill4(int *p, int base) {
  for (int i = 0; i < 4; i++)
    p[i] = base + i;
}

/* Turns enough to be worth counting and too many to be worth writing out.  The
   bound is a constant, so the count is there to be had; what declines this is
   the budget, and the loop must come back a loop. */
int total64(const int *p) {
  int s = 0;
  for (int i = 0; i < 64; i++)
    s += p[i];
  return s;
}

/* A body big enough that four copies of it cost more than the branch they
   save.  The count is four and the budget is what refuses it. */
int wide4(const int *p, int k) {
  int s = 0;
  for (int i = 0; i < 4; i++) {
    int a = p[i] * k;
    int b = a + p[i];
    int c = b * b;
    int d = c - a;
    int e = d * k;
    int f = e + c;
    int g = f * f;
    int h = g - e;
    s += h;
  }
  return s;
}

/* A loop the body of which decides something.  This one is countable and it is
   not in one block when unrolling first looks at it — if-conversion stands
   below unrolling in the order, so the two sides become selects on the round
   after, and the round after that has a single block to write out.  Six turns
   of it come to more than the budget, so what actually refuses it is the size;
   raise the budget and it is written out.  Both facts are worth having in one
   function: the shape a pass sees depends on which round it is. */
int alternating(const int *p) {
  int s = 0;
  for (int i = 0; i < 6; i++) {
    if (p[i] > 0)
      s += p[i];
    else
      s -= p[i];
  }
  return s;
}

/* A loop whose turns depend on what it read, so no arithmetic in it settles
   how many there are. */
int until_zero(const int *p) {
  int s = 0;
  int i = 0;
  while (p[i] != 0) {
    s += p[i];
    i++;
  }
  return s;
}

/* Two counted loops one inside the other.  The inner one is written out — the
   loops come innermost first — and the outer one then has a body three times
   the size, which is what puts it over the budget.  So this comes back as one
   loop of three turns around three copies of the inner body, which is the
   trade the budget exists to make: the turns that are cheap to write out are
   written out and the ones that are not are left as a loop. */
int grid(const int *p) {
  int s = 0;
  for (int i = 0; i < 3; i++)
    for (int j = 0; j < 3; j++)
      s += p[i * 3 + j];
  return s;
}

/* A counted loop that leaves early.  The count is there to be had, but the
   loop keeps two blocks however cheap it is — the block that returns cannot be
   merged into the one that tests, and if-conversion has nothing to make a
   select of where one side leaves the function.  So this is the shape the pass
   declines for being more than one block, and it is the common one: an early
   exit is what most counted loops in real code have. */
int first_negative(const int *p) {
  for (int i = 0; i < 4; i++)
    if (p[i] < 0)
      return i;
  return -1;
}
